File: python/day5.py
from utilities import read_lines


binary_values = {
    'F': '0',
    'B': '1',
    'L': '0',
    'R': '1'
}


def convert_id(boarding_pass):
    for key, value in binary_values.items():
        boarding_pass = boarding_pass.replace(key, value)
    row = int(boarding_pass[:7], 2)
    col = int(boarding_pass[7:], 2)
    return row * 8 + col


def part_1(ids):
    return max(ids)


def part_2(ids):
    ids = sorted(ids)

    for i in range(len(ids) - 1):
        if ids[i] + 1 != ids[i+1]:
            return ids[i] + 1

    raise Exception("This is unreachable")


if __name__ == '__main__':
    print('Day 5')
    data = read_lines('../inputs/day5.txt')
    ids = list(map(convert_id, data))
    print('\tPart 1: {}'.format(part_1(ids)))
    print('\tPart 2: {}'.format(part_2(ids)))

File: rust/src/day5.rs
use crate::utilities::read_lines;

fn string_to_int(s: &str) -> i32 {
    let mut ans = 0;
    for ch in s.chars() {
        ans = ans << 1;
        if ch == 'B' || ch == 'R' {
            ans += 1;
        }
    }
    ans
}

fn convert_id(s: String) -> i32 {
    let s_ = s.as_str();
    string_to_int(&s_[..7]) * 8 + string_to_int(&s_[7..])
}

fn part_1(ids: &Vec<i32>) {
    println!("\tPart 1: {}", ids[ids.len() - 1]);
}

fn part_2(ids: &Vec<i32>) {
    println!(
        "\tPart 2: {}",
        ids.windows(2).find(|pair| pair[0] + 1 != pair[1]).unwrap()[0] + 1
    );
}

pub fn main() {
    println!("Day 5");
    let mut ids: Vec<i32> = read_lines("../inputs/day5.txt")
        .into_iter()
        .map(convert_id)
        .collect();
    ids.sort();
    part_1(&ids);
    part_2(&ids);
}
